daft/io/_generator.py

```diff
@@ -0,0 +1,99 @@
+# isort: dont-add-import: from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Callable, Iterator, List
+
+from daft.daft import Pushdowns, PyTable, ScanOperatorHandle, ScanTask
+from daft.dataframe import DataFrame
+from daft.io.scan import PartitionField, ScanOperator
+from daft.logical.builder import LogicalPlanBuilder
+from daft.logical.schema import Schema
+
+if TYPE_CHECKING:
+    from daft.table.table import Table
+
+
+def _generator_factory_function(
+    func: Callable[[int, Any], Iterator["Table"]], i: int, *args: Any
+) -> Iterator["PyTable"]:
+    for table in func(i, *args):
+        yield table._table
+
+
+def read_generator(
+    generator: Callable[[int, Any], Iterator["Table"]],
```

Comment:
what do you think instead if we make this instead an `Iterator[Callable[Iterator["Table"]]]` where pass in a generator that yields a function that when called gives us a table.

Reply:
That way we don't have to define a callable for each input partition

Reply:
Yeah that sounds good, then the num_partitions is just dependent on the size of the iterator, neat!